### PR TITLE
Documentation headings & sections

### PR DESCRIPTION
**Pull request by mcabbott**
As suggested here https://github.com/FluxML/Flux.jl/pull/2055#issuecomment-1230738013 this tries to organise the sections of the documentation a bit more logically. 

Nothing is removed. The only new content is a section listing Zygote.jl as a package which Flux re-exports.

### CODE REVIEW

File: docs/src/outputsize.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+## Model Building
```
Comment (mcabbott) [inline docs/src/outputsize.md:1]: I put `outputsize` on its own page, as I don't know what it lives with. But I'm sure nobody found it under utilities. 

This is the heading it had, I think, but clearly not quite right here.

File: docs/src/training/callbacks.md [hunk 2]
```diff
@@ -0,0 +1,77 @@
+## Callback Helpers
+
+```@docs
+Flux.throttle
+Flux.stop
+Flux.skip
+```
+
+## Patience Helpers
```
Comment (mcabbott) [inline docs/src/training/callbacks.md:9]: These were in utilities

File: docs/src/training/zygote.md [hunk 3]
```diff
@@ -0,0 +1,22 @@
+# Automatic Differentiation using Zygote.jl
+
+Flux re-exports the `gradient` from [Zygote](https://github.com/FluxML/Zygote.jl), and uses this function within [`train!`](@ref) to differentiate the model. Zygote has its own [documentation](https://fluxml.ai/Zygote.jl/dev/), in particulat listing some [limitations](https://fluxml.ai/Zygote.jl/dev/limitations/).
+
+```@docs
+Zygote.gradient
```
Comment (mcabbott) [inline docs/src/training/zygote.md:6]: This is new. 

There are several docstrings for gradient, and we should separate out the Params ones from the explicit.

File: docs/src/utilities.md [hunk 4]
```diff
@@ -1,9 +1,4 @@
-# Utility Functions
-
-Flux provides utility functions which can be used to initialize your layers
-or to regularly execute callback functions.
-
-## Layer Initialisation
+# Random Weight Initialisation
 
 Flux initialises convolutional layers and recurrent cells with `glorot_uniform` by default.
```
Comment (mcabbott) [inline docs/src/utilities.md:3]: Maybe the file ought to be renamed. Utilities has a long list of `rand` variants, which could live here or in Model Tools below?

File: docs/make.jl [hunk 5]
```diff
@@ -14,28 +14,30 @@ makedocs(
...
-         "The Julia Ecosystem" => "ecosystem.md",
-         "Utility Functions" => "utilities.md",
+         "Model Tools" => [
+             "Saving & Loading" => "saving.md",
+             "Size Propagation" => "outputsize.md",
+             "Weight Initialisation" => "utilities.md",
+         ],
          "Performance Tips" => "performance.md",
-         "Datasets" => "datasets.md",
-         "Community" => "community.md"
```
Comment (mcabbott) [inline docs/make.jl:38]: Datasets became one line in the "ecosystem" page.

Community I put on the first page, after "learning flux"

File: docs/make.jl [hunk 6]
```diff
@@ -14,28 +14,30 @@ makedocs(
...
-            "Neural Network primitives from NNlib.jl" => "models/nnlib.md",
-            "Recursive transformations from Functors.jl" => "models/functors.md"
+            "NNlib.jl" => "models/nnlib.md",
+            "Functors.jl" => "models/functors.md",
          ],
          "Handling Data" => [
-             "One-Hot Encoding with OneHotArrays.jl" => "data/onehot.md",
-             "Working with data using MLUtils.jl" => "data/mlutils.md"
+             "MLUtils.jl" => "data/mlutils.md",
+             "OneHotArrays.jl" => "data/onehot.md",
```
Comment (mcabbott) [inline docs/make.jl:26]: Here I'm trying to (1) shorten these to fit on one line, and (2) put the package Name.jl to make such pages of "foreign API" visually distinct from pages about Flux itself. 

I hope that "Working with data" is clear enough from "Handling Data" just above. The actual titles on the pages remain as before.

File: docs/make.jl [hunk 7]
```diff
@@ -14,28 +14,30 @@ makedocs(
...
             "Basics" => "models/basics.md",
             "Recurrence" => "models/recurrence.md",
-            "Model Reference" => "models/layers.md",
+            "Layer Reference" => "models/layers.md",
             "Loss Functions" => "models/losses.md",
             "Regularisation" => "models/regularisation.md",
             "Advanced Model Building" => "models/advanced.md",
-            "Neural Network primitives from NNlib.jl" => "models/nnlib.md",
-            "Recursive transformations from Functors.jl" => "models/functors.md"
+            "NNlib.jl" => "models/nnlib.md",
```
Comment (mcabbott) [inline docs/make.jl:21]: I wonder if "Activation Functions" should be split out of this to be its own page.

File: docs/make.jl [hunk 8]
```diff
@@ -1,41 +1,44 @@
...
         "Home" => "index.md",
         "Building Models" => [
             "Overview" => "models/overview.md",
             "Basics" => "models/basics.md",
             "Recurrence" => "models/recurrence.md",
-            "Model Reference" => "models/layers.md",
+            "Layer Reference" => "models/layers.md",
             "Loss Functions" => "models/losses.md",
             "Regularisation" => "models/regularisation.md",
             "Advanced Model Building" => "models/advanced.md",
```
Comment (mcabbott) [inline docs/make.jl:20]: Maybe "Advanced Model Building" should be more like "Custom Layers"? That's what it describes. 

It also has a section "[Freezing Layer Parameters](https://fluxml.ai/Flux.jl/stable/models/advanced/#Freezing-Layer-Parameters)" which I think should move to the Training heading below. (And will be scrapped when Params is removed.)

File: docs/make.jl (hunk 6 above)
Comment (mcabbott) [reply docs/make.jl:27]: New on the right:
Screenshot 2022-08-29 at 19 26 02

File: docs/make.jl (hunk 7 above)
Comment (mcabbott) [reply docs/make.jl:21]: Now done

File: docs/src/training/zygote.md (hunk 3 above)
Comment (mcabbott) [reply docs/src/training/zygote.md]: `gradient` has two docstrings but one signature. Not sure how to get the right ones here, on preview I only get the `gradient(f, args...)` "explicit" one. 
```
julia> methods(gradient)
# 1 method for generic function "gradient" from Zygote:
 [1] gradient(f, args...)
     @ ~/.julia/packages/Zygote/qGFGD/src/compiler/interface.jl:95

help?> gradient
search: gradient withgradient GeneralizedEigen generate_well_conditioned_matrix

  gradient(f, args...)

  Returns a tuple containing ∂f/∂x for each argument x, the derivative (for scalar x) or the
  gradient.

  f(args...) must be a real number, see jacobian for array output.

...
────────────────────────────────────────────────────────────────────────────────────────────────

  gradient(() -> loss(), ps::Params) -> Grads

  Gradient with implicit parameters. Takes a zero-argument function, and returns a dictionary-like
  container, whose keys are arrays x in ps.
```

File: docs/make.jl (hunk 7 above)
Comment (mcabbott) [reply docs/make.jl:21]: (I believe Optimisers.jl may want a similar split, eventually, for its long lsit of rules, vs. its short list of "interface" functions.)

File: docs/src/training/zygote.md [hunk 9]
```diff
@@ -0,0 +1,39 @@
+# Automatic Differentiation using Zygote.jl
+
+Flux re-exports the `gradient` from [Zygote](https://github.com/FluxML/Zygote.jl), and uses this function within [`train!`](@ref) to differentiate the model. Zygote has its own [documentation](https://fluxml.ai/Zygote.jl/dev/), in particulat listing some [limitations](https://fluxml.ai/Zygote.jl/dev/limitations/).
+
+### Implicit style
+
+Flux uses primarily what Zygote calls "implicit" gradients, [described here](https://fluxml.ai/Zygote.jl/dev/#Explicit-and-Implicit-Parameters-1) in its documentation. 
+
+```@docs
+Zygote.gradient(f, pars::Zygote.Params)
```
Comment (Saransh-cpp) [inline docs/src/training/zygote.md:10]:
```suggestion
Zygote.gradient
```
`Zygote.gradient` should render the `gradient(() -> loss(), ps::Params) -> Grads` docstring

**Comment by Saransh-cpp** [comment]
I had something similar planned in my proposal, and this PR does all of it (even more than the proposed plan) 😅 

If it helps, this was the structure I had in mind when I was submitting a proposal - https://whimsical.com/documentation-A3YtpKeqH6ZHopZYNdgBLc

(Note: I was unsure about how the API references and re-exported packages should be placed, and I was very new to the FluxML ecosystem back then.

For instance I imagined the datasets to live under a "Data" heading, along with OneHotArrays and utility functions, but the datasets look good in the ecosystem page too!)

**Comment by mcabbott** [comment]
Cool. I don't think this is perfect, but it's a step towards a less random organisation... without editing much text.

I just saw https://github.com/JuliaComputing/MultiDocumenter.jl, possibly from you somewhere. That might be a nice alternative to having the complete list of NNlib functions included in Flux's docs. This complete list seems likely to go stale, better to have only one, inside NNlib, etc. In which case I'm not sure what happens to these sections --- having activation functions listed here makes some sense, deserves a page still? But should the section titled NNlib.jl be deleted entirely, or become just a selection (e.g. softmax) with a note to look upward for the rest?

**Comment by Saransh-cpp** [comment]
`MultiDocumenter.jl` will definitely be very helpful in increasing the accessibility of docs. I think `NNlib.jl` should eventually become a selection, and the users could be guided to `NNlib`'s docs for more information (one click away).

**Comment by mcabbott** [comment]
Now I found your slack message. Sounds like MultiDocumenter.jl is not quite ready to for prime-time?

Maybe we should merge this & then iterate? "Model Tools" sounds a bit generic but I couldn't think of a better idea... and IMO it's useful to group things. What should happen at the intro I'm not sure, maybe we need a "Getting Started" separate from "Building Models"...

**Comment by Saransh-cpp** [comment]
> Sounds like MultiDocumenter.jl is not quite ready to for prime-time?

Not ready yet :(

> Maybe we should merge this & then iterate? "Model Tools" sounds a bit generic but I couldn't think of a better idea... and IMO it's useful to group things.

Yes! We should merge this in. This looks very good to me. We might need to take care of the failing cross-checks.

> What should happen at the intro I'm not sure, maybe we need a "Getting Started" separate from "Building Models"...

This sounds good! We can have both the pages in this way?

File: docs/src/outputsize.md (hunk 1 above)
Comment (mcabbott) [reply docs/src/outputsize.md]: Now called "shape inference"

**Comment by mcabbott** [comment]
Shall we do this? And then try to figure out how to organise the intro / getting started bit.

**Review by Saransh-cpp — APPROVED** [review]
This looks good! This does add some 404 cross-references, but majority of them do not originate from Flux. I tried pushing the fixed Flux cross-references but I don't have the rights.

## Originating from Flux
```
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "train!")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter
┌ Warning: invalid local link: unresolved path in models\functors.md        
│   link.text =
│    1-element Vector{Any}:                                                 r\1bI4T\src\Writers\HTMLWriter.jl:2081
│     "Advanced Model Building and Customisation"
│   link.url = "@ref Advanced-Model-Building-and-Customisation"
└ @ Documenter.Writers.HTMLWriter
```
## Originating from Zygote
```
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "pullback")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "jacobian")
│   link.url = "zygote.html#Zygote.jacobian-Tuple{Any, Vararg{Any, N} where N}"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "withgradient")
│   link.url = "zygote.html#Zygote.withgradient-Tuple{Any, Vararg{Any, N} where N}"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "pullback")
│   link.url = "training/@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "withjacobian")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "hessian")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "hessian_reverse")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "frule")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "NoTangent()")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "RuleConfig")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "frule")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "@scalar_rule")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in training\zygote.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "RuleConfig")
│   link.url = "@ref"
└ @ Documenter.Writers.HTMLWriter 
```
## Originating from OneHotArrays (these have been fixed but are not a part of any tagged release) 
```
┌ Warning: invalid local link: unresolved path in models\losses.md
│   link.text =
│    1-element Vector{Any}:
│    1-element Vector{Any}:
│     Markdown.Code("", "onehot")
│   link.url = "onehot.html#OneHotArrays.onehot"                            r\1bI4T\src\Writers\HTMLWriter.jl:2081
└ @ Documenter.Writers.HTMLWriter C:\Users\Saransh\.julia\packages\Documenter\1bI4T\src\Writers\HTMLWriter.jl:2081
┌ Warning: invalid local link: unresolved path in data\onehot.md
│   link.text =
│    1-element Vector{Any}:
│     Markdown.Code("", "onehotbatch")                                      r\1bI4T\src\Writers\HTMLWriter.jl:2081
│   link.url = "onehot.html#OneHotArrays.onehotbatch"
└ @ Documenter.Writers.HTMLWriter 
```

These can be fixed in another PR!

**Review by Saransh-cpp — APPROVED** [review]
(Edit: Oops, I think my review was posted twice)

**Comment by mcabbott** [comment]
I haven't thought much about the links. Are the "foreign" docstrings linking to functions not documented here, or are the links just not being translated / scoped correctly? 

Anyway maybe we merge, and then look at next steps...
